File: src/main/java/phoenix/world/genlayers/UnificationLayer.java
package phoenix.world.genlayers;

import net.minecraft.util.registry.Registry;
import net.minecraft.world.biome.Biomes;
import net.minecraft.world.gen.INoiseRandom;
import net.minecraft.world.gen.area.IArea;
import net.minecraft.world.gen.layer.traits.IAreaTransformer2;
import phoenix.init.PhoenixBiomes;

import javax.annotation.Nonnull;

public enum  UnificationLayer implements IAreaTransformer2
{
    INSTANCE;
    private static final int END_BARRENS       = Registry.BIOME.getId(Biomes.END_BARRENS);
    private static final int THE_END           = Registry.BIOME.getId(Biomes.THE_END);
    private static final int SMALL_END_ISLANDS = Registry.BIOME.getId(Biomes.SMALL_END_ISLANDS);
    private static final int UNDER             = Registry.BIOME.getId(PhoenixBiomes.UNDER.get());
    private static final int HEARTVOID         = Registry.BIOME.getId(PhoenixBiomes.HEARTVOID.get());
    @Override
    public int apply(@Nonnull INoiseRandom random, IArea area1, IArea area2, int x, int z)
    {
        int phoenix = area1.getValue(x, z);
        int vanila  = area2.getValue(x, z);
        if(vanila == THE_END || vanila == SMALL_END_ISLANDS || vanila == END_BARRENS || (phoenix != UNDER && phoenix != HEARTVOID))
        {
            return vanila;
        }
        else
        {
            return phoenix;
        }
    }

    @Override
    public int getOffsetX(int x)
    {
        return x;
    }

    @Override
    public int getOffsetZ(int z)
    {
        return z;
    }
}

File: src/main/kotlin/phoenix/world/genlayers/UnificationLayer.kt
package phoenix.world.genlayers

import net.minecraft.util.registry.Registry
import net.minecraft.world.biome.Biomes
import net.minecraft.world.gen.INoiseRandom
import net.minecraft.world.gen.area.IArea
import net.minecraft.world.gen.layer.traits.IAreaTransformer2
import phoenix.init.PhxBiomes


object UnificationLayer : IAreaTransformer2
{
    override fun apply(random: INoiseRandom, area1: IArea, area2: IArea, x: Int, z: Int): Int
    {
        val phoenix = area1.getValue(x, z)
        val vanila = area2.getValue(x, z)
        return when
        {
            vanila == END_HIGHLANDS && (phoenix == UNDER || phoenix == HEART_VOID) -> phoenix
            vanila == SMALL_END_ISLANDS && phoenix == SMALL_ISLANDS_UNDER -> phoenix
            else -> vanila
        }
    }

    override fun getOffsetX(x: Int) = x
    override fun getOffsetZ(z: Int) = z

    private val END_BARRENS: Int = Registry.BIOME.getId(Biomes.END_BARRENS)
    private val THE_END: Int = Registry.BIOME.getId(Biomes.THE_END)
    private val SMALL_END_ISLANDS: Int = Registry.BIOME.getId(Biomes.SMALL_END_ISLANDS)
    private val END_HIGHLANDS : Int = Registry.BIOME.getId(Biomes.END_HIGHLANDS)

    private val UNDER: Int = Registry.BIOME.getId(PhxBiomes.UNDER)
    private val SMALL_ISLANDS_UNDER: Int = Registry.BIOME.getId(PhxBiomes.SMALL_ISLANDS_UNDER)
    private val HEART_VOID: Int = Registry.BIOME.getId(PhxBiomes.HEARTVOID)
}